File: src/BackgroundGrid.jl
# -----
 # @ # author: Xu Ran | Bao Chenyu
 # @ # date: 2023-10-29 19:17:49
 # @ # license: MIT
 # @ # language: julia
 # @ # description: 2d background grid module
 # -----

module BackgroundGridModule

using SparseArrays;

const diag_value::Float64 = -4.0;
const offdiag_value::Float64 = 1.0;

struct UniformSquareGrid
    x_0::Float64 # left bottom x
    y_0::Float64 # left bottom y
    side_len::Float64
    n::Int64 # n nodes on each side
    boundary_values::Vector{Float64}
    delta::Float64
    x_s::Vector{Float64}
    y_s::Vector{Float64}
    unknown_index_s::Vector{Int64}
    known_index_s::Vector{Int64}
    laplacian_operator::SparseMatrixCSC{Float64, Int64}
end

function index2DToIndex1D(i::Int64, j::Int64, n::Int64)::Int64
    return (i-1) * n + j;
end

function UniformSquareGrid(xy0::Vector{Float64}, side_len::Float64, n::Int64, boundary_values)::UniformSquareGrid
    delta::Float64 = side_len / (n-1);
    boundary_values = zeros(4 * (n-1)) .+ boundary_values;
    x_s::Vector{Float64} =  LinRange(0, side_len, n) .+ xy0[1];
    y_s::Vector{Float64} =  LinRange(0, side_len, n) .+ xy0[2];
    unknown_index_s = index2DToIndex1D.(Vector(2: n-1), Vector(2: n-1)', n);
    unknown_index_s = reshape(unknown_index_s', (n-2)^2);
    # known index is the complement of unknown index
    all_index_s = index2DToIndex1D.(Vector(1: n), Vector(1: n)', n);
    all_index_s = reshape(all_index_s', n^2);
    known_index_s = setdiff(all_index_s, unknown_index_s);
    laplacian_operator::SparseMatrixCSC{Float64, Int64} = spdiagm(
        -1 => ones(n^2-1) .* offdiag_value,
        0 => ones(n^2) .* diag_value,
        1 => ones(n^2-1) .* offdiag_value,
        n => ones(n^2-n) .* offdiag_value,
        -n => ones(n^2-n) .* offdiag_value
    );
    laplacian_operator ./= delta^2;
    return UniformSquareGrid(
        xy0[1],
        xy0[2],
        side_len,
        n,
        boundary_values,
        delta,
        x_s,
        y_s,
        unknown_index_s,
        known_index_s,
        laplacian_operator
    );
end

function positionAt(i::Int64, j::Int64, grid::UniformSquareGrid)::Vector{Float64}
    return [grid.x_s[i], grid.y_s[j]];
end

function positionsAt(grid::UniformSquareGrid)::Matrix{Vector{Float64}}
    n::Int64 = grid.n;
    positions_mat::Matrix{Vector{Float64}} = positionAt.(Vector(1: n), Vector(1: n)', Ref(grid));
    return positions_mat;
end

export UniformSquareGrid, positionAt, positionsAt;

end # module BackgroundGridModule

File: PythonVersion/src/BackgroundGrid.py
'''
 # @ author: Xu Ran | Bao Chenyu
 # @ date: 2023-10-27 23:05:54
 # @ license: MIT
 # @ language: python 3
 # @ description: 2D background grid
 '''
 
import numpy as np
from scipy import sparse

diag_coeff = -4.0
offdiag_coeff = 1.0

class UniformSquareGrid:
    
    # x0: left bottom of the square
    # y0: left bottom of the square
    # edge_len: edge length of the square
    # n: number of nodes on one edge of the square
    
    def __init__(self) -> None:
        self.reset()
        pass
    
    def reset(self) -> None:
        self.x0 = 0.0
        self.y0 = 0.0
        self.edge_len = 1.0
        self.n = 2
        self.setParameters((self.x0, self.y0), self.edge_len, self.n)
        pass
    
    def setParameters(self, xy0, edge_len, n) -> None:
        self.x0, self.y0 = xy0
        self.edge_len = edge_len
        self.n = n
        self.__calParameters()
        self.__calLaplacianOperator()
        pass
    
    def position(self, i, j) -> tuple:
        x_i = self.x0 + i * self.delta
        y_j = self.y0 + j * self.delta
        return x_i, y_j
        pass
    
    def __calParameters(self) -> None:
        self.delta = self.edge_len / (self.n - 1)
        self.__getKnownIndexArray()
        self.__getUknownIndexArray()
        pass
    
    def __getKnownIndexArray(self) -> None:
        # known_index_array is a 1D array
        # the index of known_index_array is the boundary node index
        self.known_index_array = np.zeros(4 * (self.n - 1), dtype = int)
        self.known_index_array[0 : self.n] = np.arange(0, self.n)
        for i_row in range(1, self.n - 1):
            self.known_index_array[(i_row-1)*2 + self.n] = i_row * self.n
            self.known_index_array[(i_row-1)*2 + self.n + 1] = (i_row + 1) * self.n - 1
            pass
        self.known_index_array[-self.n : ] = np.arange((self.n-1)*self.n, self.n*self.n)
        pass
    
    def __getUknownIndexArray(self) -> None:
        self.unknown_index_array = np.setdiff1d(np.arange(0, self.n*self.n), self.known_index_array)
        pass
    
    def __calLaplacianOperator(self) -> None:
        # Laplacian operator
        # use sparse matrix to save memory
        # the matrix is a square matrix with shape (n*n, n*n)
        # 1. diag part
        diag = np.zeros(self.n*self.n) + diag_coeff / self.delta**2
        diag_row = np.arange(0, self.n*self.n)
        diag_col = np.arange(0, self.n*self.n)
        # 2. upper neighbour to diag
        upper_neighbour = np.zeros(self.n*self.n - 1) + offdiag_coeff / self.delta**2
        upper_neighbour_row = np.arange(0, self.n*self.n - 1)
        upper_neighbour_col = np.arange(1, self.n*self.n)
        # 3. lower neighbour to diag
        lower_neighbour = np.zeros(self.n*self.n - 1) + offdiag_coeff / self.delta**2
        lower_neighbour_row = np.arange(1, self.n*self.n)
        lower_neighbour_col = np.arange(0, self.n*self.n - 1)
        # 4. upper far to diag
        upper_far = np.zeros(self.n*self.n - self.n) + offdiag_coeff / self.delta**2
        upper_far_row = np.arange(0, self.n*self.n - self.n)
        upper_far_col = np.arange(self.n, self.n*self.n)
        # 5. lower far to diag
        lower_far = np.zeros(self.n*self.n - self.n) + offdiag_coeff / self.delta**2
        lower_far_row = np.arange(self.n, self.n*self.n)
        lower_far_col = np.arange(0, self.n*self.n - self.n)
        # 6. assemble
        data = np.concatenate((diag, upper_neighbour, lower_neighbour, upper_far, lower_far))
        row = np.concatenate((diag_row, upper_neighbour_row, lower_neighbour_row, upper_far_row, lower_far_row))
        col = np.concatenate((diag_col, upper_neighbour_col, lower_neighbour_col, upper_far_col, lower_far_col))
        # 7. generate sparse matrix
        self.laplacian_operator = sparse.coo_matrix((data, (row, col)), shape=(self.n*self.n, self.n*self.n)).tocsr()
        pass
    
    pass